### PR TITLE
Fix Issue 21285 - Delegate covariance broken between 2.092 and 2.094 (git master).

### PR DESCRIPTION
**Pull request by Biotronic**

### CODE REVIEW

**Comment by UplinkCoder** [comment]
Thanks for fixing this so fast.

**Comment by Biotronic** [comment]
My pleasure - it was nagging me since I first saw the post by H. S. Teoh on the forum. I worry though that there are other cases that aren't considered here. The error/no error split feels ok, but special casing functions and delegates feels like a hack.

**Comment by UplinkCoder** [comment]
It certainly is a hack, but so is the rest.

**Comment by looked-at-me** [comment]
I would add a static assert to also ensure that the appropriate type is used, or just change the "auto" to the explicit type. Maybe do the same for the original pull request as well.

**Comment by kinke** [comment]
> It certainly is a hack, but so is the rest.

That's no excuse for adding yet another hack. Getting to the bottom of why `castTo()` fails in this case would be much appreciated.

File: test/compilable/b21285.d [hunk 1]
```diff
@@ -0,0 +1,20 @@
+// Issue 21285 - Delegate covariance broken between 2.092 and 2.094 (git master).
```
Comment (BorisCarvajal) [inline test/compilable/b21285.d:1]: Should have?
`// REQUIRED_ARGS: -unittest`

Comment (Biotronic) [reply test/compilable/b21285.d:2]: You're probably right. Sorry about this taking a while.

**Comment by Biotronic** [comment]
The static assert idea has merit, but apparently [mentioning the correct types before array creation](https://run.dlang.io/is/S8RkDP) (not *using* the types, mind you - just declaring aliases) is enough to make the code compile...

Putting the aliases after array creation makes it fail correctly, though.

**Comment by quickfur** [comment]
ping

Any updates on this? This regression is making it impossible to work on one of my projects that depends on this.

**Review by thewilsonator — APPROVED** [review]

**Comment by quickfur** [comment]
This fix is incomplete. The following code still breaks:
````
int f(string s) { throw new Exception(""); }
void main()
{
    string path;
    int bank, preset;
    void delegate(string value)[string] aa = [
        "path": (string arg) {
            path = arg;
        },
        "bank": (string arg) {
            bank = f(arg);
        },
        "preset": (string arg) {
            preset = f(arg);
        },
    ];
}
````

**Comment by quickfur** [comment]
Bug reopened: https://issues.dlang.org/show_bug.cgi?id=21285

**Pull request merged**